### PR TITLE
#310 Added already_evaluated method in Individual

### PR DESCRIPTION
**Pull request by alimustafashah**

### CODE REVIEW

**Review by HenrikMettler — CHANGES_REQUESTED** [review]
Nice! Thanks for working on this & contributing! :tada: 
Two things: 
1) I am not fully sure whether the function should return
 ```python
 not (self._fitness[self._objective_idx] is None)
```
 or 
```python 
not (any (objective_fitness is None for objective_fitness in self._fitness))
```
ie. whether to only check for the current objective or for all. Could you write a test, that covers this difference and clearly shows why to use the respective version? 

2) Could you implement the change in all occurences of the code? Eg. all examples?

**Comment by alimustafashah** [comment]
The answer to why I added **not**:-
The fitness_is_None() method is as follows:
```
def fitness_is_None(self) -> bool:
        return self._fitness[self._objective_idx] is None
```
And the initial objective() method was as follows with a **not** in the **if**:
```
def objective(individual):
    if **not individual.fitness_is_None()**:
        return individual
    [...]
```
You wanted the above **if** to be evaluated without **not**:
```
def objective(individual):
    if **individual.already_evaluated()**:
        return individual
    [...]
```
Therefore, I copied fitness_is_None() method named it already_evaluated() and added a **not** in the return statement.


Yes, I can implement the change in all occurrences of the code. Should I remove fitness_is_None method?

**Comment by alimustafashah** [comment]
The answer to why I added **not**:-
The fitness_is_None() method is as follows:
```
def fitness_is_None(self) -> bool:
        return self._fitness[self._objective_idx] is None
```
And the initial objective() method was as follows with a **not** in the **if**:
```
def objective(individual):
    if not individual.fitness_is_None():
        return individual
    [...]
```
You wanted the above **if** to be evaluated without **not**:
```
def objective(individual):
    if individual.already_evaluated():
        return individual
    [...]
```
Therefore, I copied fitness_is_None() method named it already_evaluated() and added a **not** in the return statement.


Yes, I can implement the change in all occurrences of the code. Should I remove fitness_is_None method?

**Comment by jakobj** [comment]
I agree @alimustafashah, your interpretation sounds correct. We only want to check whether the individual has been evaluated on the _current_ objective, not on any objective. If we would do the latter, we would run into issues with hurdles. @HenrikMettler what do you think, should we rename the function to reflect this and avoid all confusion?

I would suggest to adapt all tests and examples, and to keep the old method, adding a deprecation warning.

File: cgp/individual.py [hunk 1]
```diff
@@ -85,6 +85,9 @@ def _copy_user_defined_attributes(self, other: "IndividualBase") -> None:
     def fitness_is_None(self) -> bool:
         return self._fitness[self._objective_idx] is None
```
Comment (jakobj) [inline cgp/individual.py:86]: please add a deprecation warning here.

**Review by jakobj — CHANGES_REQUESTED** [review]
great, thanks for contributing @alimustafashah !

as already mentioned above, please add a deprecation warning and adapt the tests as well.

File: cgp/individual.py (hunk 1 above)
Comment (alimustafashah) [reply cgp/individual.py:86]: Okay : )

**Comment by alimustafashah** [comment]
Is it good now? Now, I should replace all instances of fitness_is_None() with already_evaluated() method, right?

File: cgp/individual.py [hunk 2]
```diff
@@ -83,8 +84,15 @@ def _copy_user_defined_attributes(self, other: "IndividualBase") -> None:
                 setattr(other, attr, copy.deepcopy(getattr(self, attr)))
 
     def fitness_is_None(self) -> bool:
+        warnings.warn(
+            "fitness_is_None is deprecated, use already_evaluated instead"
+            , DeprecationWarning
```
Comment (jakobj) [inline cgp/individual.py:89]:
```suggestion
            "`fitness_is_None` is deprecated and will be remobed in the 0.4 release. Please use `already_evaluated` instead"
            , DeprecationWarning
```

**Review by jakobj — COMMENTED** [review]
added a small suggestion for improving the warning message.

yes, now replace all use of `fitness_is_None` with your new method. i will also enable CI.

**Comment by jakobj** [comment]
i think there is an issue with code formatting @alimustafashah, please have a look at the output from the github actions workflow. you can fix them by running black, mypy, isort, flake8 locally.

@mschmidt87 did we document somewhere how the code formatting should be/can be done locally (beside git commit hooks ;) )?

**Comment by mschmidt87** [comment]
Why "beside git commit hooks"? Please use pre-commit hooks. To set them up, just do:
```
pip install -f extra-requirements.txt
pre-commit install
```
and then commit your files. It is really just as simple as that and will save work/time/troubles.

**Comment by alimustafashah** [comment]
> pip install -f extra-requirements.txt

I am getting the following error on running the command
```
pip install -f extra-requirements.txt
ERROR: You must give at least one requirement to install (maybe you meant "pip install extra-requirements.txt"?)
```

**Comment by HenrikMettler** [comment]
@alimustafashah can you try: `pip install -r extra-requirements.txt `?

**Comment by alimustafashah** [comment]
> @alimustafashah can you try: `pip install -r extra-requirements.txt `?

Got this:-
```
pip install -r extra-requirements.txt
Collecting matplotlib~=3.2.1
  Downloading matplotlib-3.2.2-cp38-cp38-win32.whl (9.0 MB)
     |████████████████████████████████| 9.0 MB 595 kB/s
Requirement already satisfied: scipy~=1.4.1 in c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\site-packages (from -r extra-requirements.txt (line 3)) (1.4.1)
Collecting sympy~=1.6.1
  Downloading sympy-1.6.2-py3-none-any.whl (5.8 MB)
     |████████████████████████████████| 5.8 MB 298 kB/s
ERROR: Could not find a version that satisfies the requirement torch~=1.5.0 (from versions: 0.1.2, 0.1.2.post1, 0.1.2.post2)
ERROR: No matching distribution found for torch~=1.5.0
```

**Comment by HenrikMettler** [comment]
This seems to be an issue with installing torch. Which python version are you using?

**Comment by alimustafashah** [comment]
> This seems to be an issue with installing torch. Which python version are you using?

Python 3.8.1

**Comment by HenrikMettler** [comment]
what happens if you just do `pip install torch` ? 
Otherwise you can just remove the line `torch ~= 1.5.0` from extra_requirements.txt

**Comment by alimustafashah** [comment]
> what happens if you just do `pip install torch` ?
> Otherwise you can just remove the line `torch ~= 1.5.0` from extra_requirements.txt

```
pip install torch
Collecting torch
  Downloading torch-0.1.2.post2.tar.gz (128 kB)
     |████████████████████████████████| 128 kB 504 kB/s
Requirement already satisfied: pyyaml in c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\site-packages (from torch) (5.4.1)
WARNING: The candidate selected for download or install is a yanked version: 'torch' candidate (version 0.1.2.post2 at https://files.pythonhosted.org/packages/f8/02/880b468bd382dc79896eaecbeb8ce95e9c4b99a24902874a2cef0b562cea/torch-0.1.2.post2.tar.gz#sha256=a43e37f8f927c5b18f80cd163daaf6a1920edafcab5102e02e3e14bb97d9c874 (from https://pypi.org/simple/torch/))
Reason for being yanked: 0.1.2 is past it's support date and confuses users on unsupported platforms
Using legacy 'setup.py install' for torch, since package 'wheel' is not installed.
Installing collected packages: torch
    Running setup.py install for torch ... error
    ERROR: Command errored out with exit status 1:
     command: 'c:\users\ali mustafa\appdata\local\programs\python\python38-32\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Ali Mustafa\\AppData\\Local\\Temp\\pip-install-co0jj_92\\torch_e2066bf8141443d9943bbe79b62ebf0d\\setup.py'"'"'; __file__='"'"'C:\\Users\\Ali Mustafa\\AppData\\Local\\Temp\\pip-install-co0jj_92\\torch_e2066bf8141443d9943bbe79b62ebf0d\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Ali Mustafa\AppData\Local\Temp\pip-record-0ees6cuh\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\ali mustafa\appdata\local\programs\python\python38-32\Include\torch'
         cwd: C:\Users\Ali Mustafa\AppData\Local\Temp\pip-install-co0jj_92\torch_e2066bf8141443d9943bbe79b62ebf0d\
    Complete output (23 lines):
    running install
    running build_deps
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\Ali Mustafa\AppData\Local\Temp\pip-install-co0jj_92\torch_e2066bf8141443d9943bbe79b62ebf0d\setup.py", line 225, in <module>
        setup(name="torch", version="0.1.2.post2",
      File "c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\site-packages\setuptools\__init__.py", line 145, in setup
        return distutils.core.setup(**attrs)
      File "c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\distutils\dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\distutils\dist.py", line 985, in run_command
        cmd_obj.run()
      File "C:\Users\Ali Mustafa\AppData\Local\Temp\pip-install-co0jj_92\torch_e2066bf8141443d9943bbe79b62ebf0d\setup.py", line 99, in run
        self.run_command('build_deps')
      File "c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "c:\users\ali mustafa\appdata\local\programs\python\python38-32\lib\distutils\dist.py", line 985, in run_command
        cmd_obj.run()
      File "C:\Users\Ali Mustafa\AppData\Local\Temp\pip-install-co0jj_92\torch_e2066bf8141443d9943bbe79b62ebf0d\setup.py", line 51, in run
        from tools.nnwrap import generate_wrappers as generate_nn_wrappers
    ModuleNotFoundError: No module named 'tools.nnwrap'
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'c:\users\ali mustafa\appdata\local\programs\python\python38-32\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Ali Mustafa\\AppData\\Local\\Temp\\pip-install-co0jj_92\\torch_e2066bf8141443d9943bbe79b62ebf0d\\setup.py'"'"'; __file__='"'"'C:\\Users\\Ali Mustafa\\AppData\\Local\\Temp\\pip-install-co0jj_92\\torch_e2066bf8141443d9943bbe79b62ebf0d\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Ali Mustafa\AppData\Local\Temp\pip-record-0ees6cuh\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\ali mustafa\appdata\local\programs\python\python38-32\Include\torch' Check the logs for full command output.
```

**Comment by jakobj** [comment]
hmm, this may be a  windows specific issue. you should check an installation guide for torch on windows @alimustafashah

**Comment by alimustafashah** [comment]
> @alimustafashah any luck fixing the torch installation? Or did you try installing the extra-requirements w/o torch?

I have tried it on Linux but still there is error. I have attached photo of the terminal, please have a look.

terminal

**Comment by mschmidt87** [comment]
The correct command is `pip install -r extra-requirements.txt` (see here: https://pip.pypa.io/en/latest/user_guide/#requirements-files). It doesn't work for you because it can't find the file. Apparently you are not in the correct folder. Please navigate to the root folder of the repository and repeat the last command you executed above (in your screenshot).

**Comment by alimustafashah** [comment]
> The correct command is `pip install -r extra-requirements.txt` (see here: https://pip.pypa.io/en/latest/user_guide/#requirements-files). It doesn't work for you because it can't find the file. Apparently you are not in the correct folder. Please navigate to the root folder of the repository and repeat the last command you executed above (in your screenshot).

It was named requirements.txt and I installed it. Following is the solution:-
terminal

Its solved but now I am installing pre-commit and the system is giving me a warning. It is saying me if I want to proceed I should add classic flag. Should I proceed with classic flag? Following is the screenshot:-
teminal2

**Comment by jakobj** [comment]
> > The correct command is `pip install -r extra-requirements.txt` (see here: https://pip.pypa.io/en/latest/user_guide/#requirements-files). It doesn't work for you because it can't find the file. Apparently you are not in the correct folder. Please navigate to the root folder of the repository and repeat the last command you executed above (in your screenshot).
> 
> It was named requirements.txt and I installed it. Following is the solution:-

yes, this install part of the required packages. please execute the same command but with `requirements.txt` -> `extra_requirements.txt`.
also, you might want to look into using pip without sudo, there are several guides on this available via the search engine of your choice.

> 
> Its solved but now I am installing pre-commit and the system is giving me a warning. It is saying me if I want to proceed I should add classic flag. Should I proceed with classic flag? Following is the screenshot:-

have you tried install `pre-commit` via pip? it's a python packages afaik. you could also try with the `classic` flag, but only if you know what you are doing.

**Comment by jakobj** [comment]
closing due to inactivity.

**Pull request closed without merging**